### PR TITLE
Expose config option to enable low-level S3 HTTP operation logging.

### PR DESCRIPTION
**Pull request by seizethedave**
#### What this PR does

Minio has support for tracing all object store HTTP request and response metadata to the log. We recently had some operational research where this would have been useful to see. Add an `s3.trace.enabled` boolean config flag to turn this on.

This config option can also be used in `mimirtool`'s bucket validator command.

[Here's](https://github.com/grafana/mimir/blob/2bdf8128c5134432977c0e038165458f12fed9c2/vendor/github.com/minio/minio-go/v7/api.go#L479) where Minio writes these logs. They're written in this format, roughly:

```
2024-07-15 20:50:11.302	ts=2024-07-15T20:50:11.294149396Z caller=stdlib.go:105 level=debug s3TraceMsg=---------START-HTTP---------
2024-07-15 20:50:11.302	ts=2024-07-15T20:50:11.294340157Z caller=stdlib.go:105 level=debug s3TraceMsg="HEAD /bucketname/markers/tenant-deletion-mark.json HTTP/1.1\r\n {{request headers}}"
2024-07-15 20:50:11.302	ts=2024-07-15T20:50:11.2943726Z caller=stdlib.go:105 level=debug s3TraceMsg="HTTP/1.1 404 Not Found\r\nConnection: close\r\nContent-Type: application/xml\r\nDate: Mon, 15 Jul 2024 20:50:10 GMT\r\nServer: AmazonS3\r\nX-Amz-Id-2: foo\r\nX-Amz-Request-Id: bar\r"
2024-07-15 20:50:11.302	ts=2024-07-15T20:50:11.294383061Z caller=stdlib.go:105 level=debug s3TraceMsg=---------END-HTTP---------
```

#### Checklist

- [x] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.

### CODE REVIEW

File: docs/sources/mimir/configure/configuration-parameters/index.md [hunk 1]
```diff
@@ -5141,6 +5141,12 @@ http:
   # (advanced) Override the expected name on the server certificate.
   # CLI flag: -<prefix>.s3.http.tls-server-name
   [tls_server_name: <string> | default = ""]
+
+trace:
+  # (advanced) If enabled, low-level S3 HTTP operation info will be logged at
+  # debug level.
```
Comment (tacole02) [inline docs/sources/mimir/configure/configuration-parameters/index.md:5147]:
```suggestion
  # (advanced) If enabled, low-level S3 HTTP operation information is logged at the
  # debug level.
```

Comment (tacole02) [reply docs/sources/mimir/configure/configuration-parameters/index.md:5147]: We try to avoid using "will" in the docs.

**Review by tacole02 — APPROVED** [review]
Looks good! Thanks for updating :)

**Review by charleskorn — APPROVED** [review]

File: pkg/storage/bucket/s3/config.go [hunk 2]
```diff
@@ -269,6 +271,14 @@ func parseKMSEncryptionContext(data string) (map[string]string, error) {
 	return decoded, err
 }
 
+type TraceConfig struct {
+	Enable bool `yaml:"enable" category:"advanced"`
+}
+
+func (cfg *TraceConfig) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
+	f.BoolVar(&cfg.Enable, prefix+"enable", false, "If enabled, low-level S3 HTTP operation information is logged at the debug level.")
```
Comment (narqo) [inline pkg/storage/bucket/s3/config.go:279]: (nit) I recall @pstibrany mentioned, that Mimir tries to stick to "enabled" for on/off settings.

```suggestion
	f.BoolVar(&cfg.Enable, prefix+"enabled", false, "If enabled, low-level S3 HTTP operation information is logged at the debug level.")
```

And the same for the fields in the config itself.

Comment (seizethedave) [reply pkg/storage/bucket/s3/config.go:279]: Ah, thanks for that. I'll take a look at that this morning.

**Pull request merged**